Experimental HDR support on macOS using colorSpace approach

anastasiuspernat opened the pull request:
Enabling support for HDR monitors (in manual mode).

Main discussion is here: https://github.com/mpv-player/mpv/issues/7341

Added a new switch called `--macos_hdr_csp` (similar to d3d color space approach at https://github.com/mpv-player/mpv/pull/5804). It specifies the extended HDR color space. Doesn't require any shader or video render pipeline modification. Auto-activates EDR/HDR mode on the video layer.

Currently supported color spaces (passed to `--macos_hdr_csp` switch):
- `displayP3_PQ_EOTF` DCI P3 primaries, PQ transfer function - that's the most common color profile for most videos
- `displayP3_HLG` DCI P3 primaries, and the HLG gamma
- `extendedLinearDisplayP3` P3 primaries and linear transfer function
- `itur_2020` BT.2020 
- `itur_2020_PQ_EOTF` BT.2020 with PQ transfer function

This can (and probably must) be set together with `target-trc` switch. Both values must correspond to video's color primaries/transfer characteristics in order to get the correct color reproduction.

Tested on varios video footage on macOS Catalina 10.15.5. Produces same results as QuickTime player on the same footage.

Akemi commented on video/out/cocoa_cb_common.swift:
```diff
@@ -122,7 +122,41 @@ class CocoaCB: Common {
+            var name = "" as CFString
+
+            // HDR is supported fully only on macOS 10.15 and higher
+#if HAVE_MACOS_10_15_FEATURES
```
for HAVE_MACOS_10_15_FEATURES to work you need to add a new configure flag for it to exists. otherwise this won't work. see https://github.com/mpv-player/mpv/blob/master/wscript#L816-L831

Akemi commented on video/out/cocoa_cb_common.swift:
```diff
@@ -122,7 +122,41 @@ class CocoaCB: Common {
+#if HAVE_MACOS_10_15_FEATURES
+                switch (libmpv.macOpts.macos_hdr_csp)
+                {
+                    case 0: name = CGColorSpace.displayP3_HLG
```
macOS 10.14.6 guard

Akemi commented on video/out/cocoa_cb_common.swift:
```diff
@@ -122,7 +122,41 @@ class CocoaCB: Common {
+                {
+                    case 0: name = CGColorSpace.displayP3_HLG
+                            break;
+                    case 1: name = CGColorSpace.displayP3_PQ_EOTF
```
macOS 10.14.6 guard

Akemi commented on video/out/cocoa_cb_common.swift:
```diff
@@ -122,7 +122,41 @@ class CocoaCB: Common {
+                            break;
+                    case 1: name = CGColorSpace.displayP3_PQ_EOTF
+                            break;
+                    case 2: name = CGColorSpace.extendedLinearDisplayP3
```
macOS 10.14.3 guard

Akemi commented on video/out/cocoa_cb_common.swift:
```diff
@@ -122,7 +122,41 @@ class CocoaCB: Common {
+                            break;
+                    case 2: name = CGColorSpace.extendedLinearDisplayP3
+                            break;
+                    case 3: name = CGColorSpace.itur_2020
```
macOS 10.11 guard

Akemi commented on video/out/cocoa_cb_common.swift:
```diff
@@ -122,7 +122,41 @@ class CocoaCB: Common {
+                                name = CGColorSpace.itur_2020_HLG 
+                            }
+                            break;
+                    case 5: name = CGColorSpace.itur_2020_PQ_EOTF
```
macOS 10.14.6 guard

Akemi commented on video/out/cocoa_cb_common.swift:
```diff
@@ -122,7 +122,41 @@ class CocoaCB: Common {
+            if (name as String != "")
+            {
+                layer?.colorspace = CGColorSpace(name: name)
+            } else
```
i am not a big fan of the if else here. i think it would be the best to create a variable at the beginning with the auto/default colorSpace, overwrite it with the switch case and set this variable to the layer's color space.

Akemi commented on osdep/macosx_application.m:
```diff
@@ -77,7 +82,9 @@
-        .cocoa_cb_10bit_context = 1
+        .cocoa_cb_10bit_context = 1,
+        // Experimental #HDR on #macOS        
+        .macos_hdr_csp = -1
```
there is no -1 in the definition of the option. so you shouldn't se this value here or you need to add it to the option choices. i would add an "auto" option and make this -1.

anastasiuspernat replied:
I will add a "none" option, as "auto" should be probably reserved for future functionality where color profile is taken from video header and auto-assigned. How does that sound?

Akemi replied:
none would be wrong though, since it should be the same as the display in that case. also from the doku of the d3d11 omplementation ""auto" is the default, which will select the color space of the desktop on which the swap chain is located.". it should be consistent.

anastasiuspernat replied:
OK, got it.

Akemi commented on video/out/mac/gl_layer.swift:
```diff
@@ -109,6 +109,13 @@ class GLLayer: CAOpenGLLayer {
+        // Experimental #HDR on #macOS        
+        // This must be enabled to reproduce HDR content 
+        if (ccb.libmpv.macOpts.macos_hdr_csp != -1)
+        {
```
opening curly braces should be on the line as the if

Akemi commented on video/out/mac/view.swift:
```diff
@@ -33,6 +33,12 @@ class View: NSView {
+        // Experimental #HDR on #macOS        
+        // This must be enabled to reproduce HDR content 
+        if mpv?.macOpts.macos_hdr_csp != -1
+        {
```
opening curly braces should be on the line as the if

Akemi commented on video/out/cocoa_cb_common.swift:
```diff
@@ -122,7 +122,41 @@ class CocoaCB: Common {
         libmpv.setRenderICCProfile(colorSpace)
         if #available(macOS 10.11, *) {
-            layer?.colorspace = colorSpace.cgColorSpace
+
```
i would most likely move the whole color space functinality added here into its own function(s) in video/out/mac/common.swift. for several reasons:
- it uses a "mac" prefixed option that should be available for all common mac code, not just the gl_layer and cocoa_cb
- makes it reusable
- easier to maintain

probably best to implement part of this functionality here https://github.com/mpv-player/mpv/blob/master/video/out/mac/common.swift#L397-L399 and call the super.updateICCProfile() in the cocoa-cb specific part.

anastasiuspernat replied:
But isn't this functionality `GLLayer`-specific (which is only seems to be available in cocoa_cb)?

Akemi replied:
the colorspace parameter is the same for the metal layer. see my moltenvk PR for example: https://github.com/mpv-player/mpv/pull/7482/files#diff-28b1978e8194b3980f86f4f9213b4c9d0416af71f8bbbe83535ecb6442b306a1R147-R149

same for wantsExtendedDynamicRangeContent  https://developer.apple.com/documentation/quartzcore/cametallayer/1478161-wantsextendeddynamicrangecontent

Akemi commented on video/out/mac/gl_layer.swift:
```diff
@@ -109,6 +109,13 @@ class GLLayer: CAOpenGLLayer {
+        // This must be enabled to reproduce HDR content 
+        if (ccb.libmpv.macOpts.macos_hdr_csp != -1)
+        {
+            wantsExtendedDynamicRangeContent = true
```
scattering this option to 3 different places (view, layer, cocoa_cb_common) makes it a bit hard to maintain i believe. 

it might be good idea to set up the whole thing in this function https://github.com/mpv-player/mpv/blob/master/video/out/mac/common.swift#L129-L141 there you have the view and the layer.
you would not need to check the option several times. though you need to type cast the CALayer to a CAOpeGLLayer in an "if let", for new layers we need to add another type casting (MetalLayer in the future)

Akemi commented:
i added a few comments. it would also be nice to have some mpv option examples and some screenshots with before, now and quicktime. mostly out of curiosity and for future reference.

Akemi commented on osdep/macosx_application.m:
```diff
@@ -68,6 +68,11 @@
             {"auto", -1}, {"no", 0}, {"yes", 1})},
         {"cocoa-cb-10bit-context", OPT_FLAG(cocoa_cb_10bit_context)},
+        // Experimental #HDR on #macOS        
+        {"macos-hdr-csp", OPT_CHOICE(macos_hdr_csp,
```
also needs to be added in our readme

anastasiuspernat commented on video/out/cocoa_cb_common.swift:
```diff
@@ -122,7 +122,41 @@ class CocoaCB: Common {
+            var name = "" as CFString
+
+            // HDR is supported fully only on macOS 10.15 and higher
+#if HAVE_MACOS_10_15_FEATURES
```
Sounds good! Will fix it everywhere.

anastasiuspernat commented:
I hope I fixed it! Updated everything by your comments and added new info to options.rst.

Also here are reference images from my tests. They don't fully show the dynamic range of the image (this part is done on the monitor side) but they do give some idea:
![HDR_mpv_illustration](https://user-images.githubusercontent.com/1848380/102841447-d1097200-43b9-11eb-8612-931f8bec42b8.jpg)

avih commented:
The commits add (and update) `.DS_Store`, is that intentional?

anastasiuspernat commented:
Sorry about `.DS_Store`s. No it's not intentional, they were auto-added by Finder. When I develop on Mac I usually add `.DS_Store` to `.gitignore`. I think I removed them all in the final commit, no?

avih commented:
> I think I removed them all in the final commit, no?

I didn't look at the commits one by one, but assuming you intend to keep the commits incremental without squashing, then removing them only at the last commit is not enough - you'll need to remove them from all commits.

anastasiuspernat commented:
No I don't intend to keep all the commits. Would be happy if you could help me to squash them the right way if that's necessary. I think only the current final stage is important, thank you.

avih commented:
Not necessary to squash them right away, depends on how you want to present it - incremental changes or one commit which changes over time. But you can do that it you prefer to work only on one commit.

This is git territory, but you can squash commits locally by running `git rebase -i` and replace `pick` with `s` (for squash) for all but the first new commit, then save the file, and if you're happy with the result - `git push -f` would update the PR with the squashed commit.

For fllowup changes, you can either do the same with new commits before you push them, or just amend the existing commit, and again force push will update the PR. Force (`-f`) is required because it rewrites the history - acceptable in a PR but generally highly discouraged elsewhere.

anastasiuspernat commented:
OK great! Thanks for the details. One commit per major followup would definitely look better. Will squash it on next iteration.

avih commented:
There are no hard rules, but the convention is that you can force push in a PR (change existing commits), but preferably keep a set of commits which is intended to be pushed individually, where each commit does one "thing". Deciding what a "thing" is is subjective, but if your code does two unrelated things for instance, then it's preferable to keep those in two distinct commits.

Akemi commented on osdep/macosx_application.m:
```diff
@@ -68,6 +68,11 @@
         {"cocoa-cb-sw-renderer", OPT_CHOICE(cocoa_cb_sw_renderer,
             {"auto", -1}, {"no", 0}, {"yes", 1})},
         {"cocoa-cb-10bit-context", OPT_FLAG(cocoa_cb_10bit_context)},
+        // Experimental #HDR on #macOS        
```
you can drop this comment.

Akemi commented on video/out/mac/common.swift:
```diff
@@ -138,6 +141,14 @@ class Common: NSObject {
         view.layerContentsPlacement = .scaleProportionallyToFit
+
+        // Experimental #HDR on #macOS        
+        // This must be enabled to reproduce HDR content 
```
drop the comment

Akemi commented on video/out/mac/common.swift:
```diff
@@ -395,7 +406,58 @@ class Common: NSObject {
+            var name = "" as CFString
+
+// HDR is supported fully only on macOS 10.15 and higher
+// some of the HDR color spaces are available on 10.14 though
```
drop the comments.

Akemi commented on video/out/mac/common.swift:
```diff
@@ -138,6 +141,14 @@ class Common: NSObject {
+        // Experimental #HDR on #macOS        
+        // This must be enabled to reproduce HDR content 
+        if (mpv?.macOpts.macos_hdr_csp != -1) {
+            (view.layer as! CAOpenGLLayer).wantsExtendedDynamicRangeContent = true
```
please no forced casts, use ? instead.

Akemi commented on osdep/macosx_application.m:
```diff
@@ -68,6 +68,11 @@
+        // Experimental #HDR on #macOS        
+        {"macos-hdr-csp", OPT_CHOICE(macos_hdr_csp,
+            {"displayP3_HLG", 0}, {"displayP3_PQ_EOTF", 1}, {"extendedLinearDisplayP3", 2},
+            {"itur_2020", 3}, {"itur_2020_HLG", 4}, {"itur_2020_PQ_EOTF", 5}
```
-1/auto still needs to be added here.

Akemi commented on video/out/mac/common.swift:
```diff
@@ -42,12 +42,15 @@ class Common: NSObject {
 
     init(_ mpLog: OpaquePointer?) {
         log = LogHelper(mpLog)
+        colorSpace = nil
```
if not set, it's nil anyway.

Akemi commented on video/out/mac/common.swift:
```diff
@@ -395,7 +406,58 @@ class Common: NSObject {
 
     func updateICCProfile() {
-        log.sendWarning("updateICCProfile not implemented")
+        guard let windowColorSpace = window?.screen?.colorSpace else {
```
the name is technically wrong. it's not the window's colorspace.

Akemi commented on video/out/mac/common.swift:
```diff
@@ -395,7 +406,58 @@ class Common: NSObject {
+
+// HDR is supported fully only on macOS 10.15 and higher
+// some of the HDR color spaces are available on 10.14 though
+#if HAVE_MACOS_10_15_FEATURES
```
that won't work. you use lower macOS features within this. like i said you should do it similar to https://github.com/mpv-player/mpv/blob/master/video/out/mac/title_bar.swift#L215-L254

Akemi commented on video/out/mac/common.swift:
```diff
@@ -395,7 +406,58 @@ class Common: NSObject {
+// HDR is supported fully only on macOS 10.15 and higher
+// some of the HDR color spaces are available on 10.14 though
+#if HAVE_MACOS_10_15_FEATURES
+                switch (mpv?.macOpts.macos_hdr_csp)
```
the indentation is wrong here.

anastasiuspernat replied:
Fixed it! Also seems that even though some color spaces are declared as supported by 10.14 SDK they are available only on 10.15+. Travis didn't let me. So put most of those inside `HAVE_MACOS_10_15_FEATURES`.

Akemi commented on video/out/mac/common.swift:
```diff
@@ -395,7 +406,58 @@ class Common: NSObject {
+                (view?.layer as! CAOpenGLLayer).colorspace = CGColorSpace(name: name)
+            } else
+            {
+                (view?.layer as! CAOpenGLLayer).colorspace = windowColorSpace.cgColorSpace
```
no forced casts. i still dislike the else case here.

anastasiuspernat replied:
Need some help here. How else can I set color space on the `view.layer`? It's `CALayer`, right? Doesn't have `colorspace` property.

Also what's wrong with else, you mean the curly bracket?

Akemi replied:
typecasting is okay, but you should not force typecasting by any means, eg as? instead of as!. the latter case will crash when failing. see https://docs.swift.org/swift-book/LanguageGuide/TypeCasting.html

the else is unnecessary there.
```
colorSpcace = defaultColorSpace
if someCondtion == true {
    colorSpcace = differentColorSpace
} 
```

anastasiuspernat replied:
Unfortunately this way typecasting doesn't work. With as? I receive a compilation error about unwrapping CAOpenGLLayer, probably `view.layer` has a wrong type definition?

Akemi replied:
something like this should haves worked. though i would need to see the specific error.
```
guard let newlayer = layer as? CAOpenGLLayer else {
      return
 }
// access newlayer here
```

Akemi commented on video/out/mac/common.swift:
```diff
@@ -395,7 +406,58 @@ class Common: NSObject {
+                            break;
+                }
+#endif            
+            // Experimental #HDR on #macOS        
```
drop the comment.

Akemi commented:
i am not sure if what we see on those screenshots is 'intended' or more likely a 'coincidence' that it looks somewhat similar/okay'ish.

Akemi commented on video/out/mac/common.swift:
```diff
@@ -42,12 +42,15 @@ class Common: NSObject {
 
     var cursorVisibilityWanted: Bool = true
 
+    var colorSpace: NSColorSpace?
```
i believe this variable is not needed.

anastasiuspernat replied:
It's used in `updateICCProfile` for `setRenderICCProfile` in `cocoa_cb_common.swift`.

Akemi replied:
what i meant here, is that we don't need a variable that is available to the whole object of that class but just a function local variable would have done perfectly fine.

anastasiuspernat commented on DOCS/man/options.rst:
```diff
@@ -5826,6 +5826,30 @@ The following video options are currently all specific to ``--vo=gpu`` and
+        ITU-R BT.2020
+    itur_2020_PQ_EOTF 
+        ITU-R BT.2020 with PQ transfer function
+
```
From my experience those are the only ones that produce HDR colors. For the sake of testing and completeness though I can add all color spaces from `CGColorSpace`.

Akemi replied:
yes probably not all are useful, but srgb and linearsrgb might be for example.

anastasiuspernat replied:
I've added srgb and linearsrgb. Let me know if we need to add anything else.

anastasiuspernat commented on DOCS/man/options.rst:
```diff
@@ -5826,6 +5826,30 @@ The following video options are currently all specific to ``--vo=gpu`` and
     OS X only.
 
+``--macos-hdr-csp=<type>``
+    Activate experimental HDR output color space to utilize macOS X HDR display support.
```
Yeah haha, sorry. Too much Objective-C coding in good old days.

anastasiuspernat commented on DOCS/man/options.rst:
```diff
@@ -5826,6 +5826,30 @@ The following video options are currently all specific to ``--vo=gpu`` and
+    auto    
+        Select the color space of the desktop without activating HDR mode
+    displayP3_PQ_EOTF 
+        DCI P3 primaries, PQ transfer function - that's the most common color profile for most videos.
```
It's hard to argue but please note that this option should correspond to "Mastering display color primaries" parameter of the video. It's not the same as "color primaries". From my experience "color primaries" of the video doesn't provide info about HDR properties of the video stream.

anastasiuspernat commented:
> i am not sure if what we see on those screenshots is 'intended' or more likely a 'coincidence' that it looks somewhat similar/okay'ish.

They do look similar. The HDR is hardware-reproduced by monitor, there's no way on the software side to take screenshots that precisely show extended dynamic range. Probably the screenshots should be taken as pictures from the monitor using a good camera. I have one, but I don't have enough time, I'll see what I can do about it!

Also squashed the PR and updated code by our conversation. Answered on the rest of the issues.

skrew commented:
Please don't forget iOS version for the HDR support, i think is almost the same as MacOS

Akemi commented:
> Please don't forget iOS version for the HDR support, i think it's almost the same as MacOS

this has nothing todo with iOS though. the is no iOS backend in mpv.on iOS you will use libmpv and there the layers is handled by the libmpv user. so it's entirely in the hands of the one using libmpv on iOS to support this.

anastasiuspernat commented:
@Akemi 
Are there still any problems? Do I need to fix anything? Thank you and Happy New Year!

Akemi commented:
i just didn't get around looking at it again.

Akemi commented:
i opened my own PR #8485. i don't mean this to be disrespectful towards you and i hope it isn't seen as this. but what i had in mind and what you did here was a bit too far apart for me to explain everything i wanted. it would be nice if you could also test my PR.

nevertheless i will give you some more feedback on this PR for the future.

Akemi commented on osdep/macosx_application.m:
```diff
@@ -68,6 +68,11 @@
+        {"macos-output-csp", OPT_CHOICE(macos_output_csp,
+            {"auto", -1}, {"displayP3_HLG", 0}, {"displayP3_PQ_EOTF", 1}, 
+            {"extendedLinearDisplayP3", 2}, {"itur_2020", 3}, {"itur_2020_HLG", 4}, 
+            {"itur_2020_PQ_EOTF", 5}, {"extendedSRGB",6}, {"extendedLinearSRGB",7}
```
the extended color spaces expect component values below 0 and above 1.0. that's what mpv can't do atm and with that those color spaces are a bit useless atm.

Akemi commented on video/out/cocoa_cb_common.swift:
```diff
@@ -115,15 +115,8 @@ class CocoaCB: Common {
-            layer?.colorspace = colorSpace.cgColorSpace
-        }
+        super.updateICCProfile()    
+        libmpv.setRenderICCProfile(colorSpace!)
```
never use a forced unwrapping in swift if not necessary. always use a guard, if let or ? instead of !. otherwise it will only end up in hard crashes.

Akemi commented on video/out/mac/common.swift:
```diff
@@ -138,6 +140,12 @@ class Common: NSObject {
+
+        if (mpv?.macOpts.macos_output_csp != -1) {
+            (view.layer as! CAOpenGLLayer).wantsExtendedDynamicRangeContent = true
+            view.wantsExtendedDynamicRangeOpenGLSurface = true;
```
semicolon not needed

Akemi commented on video/out/mac/common.swift:
```diff
@@ -395,7 +403,61 @@ class Common: NSObject {
+            if #available(macOS 10.12, *) {
+                switch (mpv?.macOpts.macos_output_csp)
+                {
+                    case 6: name = CGColorSpace.extendedSRGB; break
```
breaks are not necessary on swift switch case. it will only fall through if an explicit fallthrough keyword is used.

anastasiuspernat commented:
> i opened my own PR #8485. i don't mean this to be disrespectful towards you and i hope it isn't seen as this. but what i had in mind and what you did here was a bit too far apart for me to explain everything i wanted. it would be nice if you could also test my PR.
> 
> nevertheless i will give you some more feedback on this PR for the future.

No worries haha! You better know your codebase, also Swift is a new language to me.

In any case my main intention was to make IINA (mpv-based video player) play with my HDR monitor on macOS. I thought mpv is the main culprit but now I see it's not. My fork of IINA now plays HDR videos perfectly so my main goal is achieved and I just wanted to give back my findings (which are based on your knowledge) to the world.

I rarely take part in opensource and it was a lot of fun. 

I'll try to test and answer when I have time! Thanks again for the guidance.

DavidSchechter commented:
@Akemi @anastasiuspernat any progress on this?

ValZapod commented:
> * DCI P3 primaries

You mean p3-d65. That is display-p3 in format and target-prim, dci-p3 uses dci whitepoint, not D65.

BTW, https://patchwork.ffmpeg.org/project/ffmpeg/patch/tencent_5C530E884D33E1B36BD5C8A9124EA5C54905@qq.com/

JulyIghor commented:
any progress on this?

Akemi commented:
sorry for stalling this for so long. i am going to close this one i favour of #8485. i will also revisit #8485 and see how we will proceed with it. also see the last comment und the linked issue for discussion (https://github.com/mpv-player/mpv/issues/7341#issuecomment-1967674781).
